src/QsCompiler/Compiler/Templates/QirDriverCpp.tt

```diff
@@ -1,16 +1,268 @@
+
+map<string, PauliId> PauliMap{
+    {"PauliI", PauliId::PauliId_I},
+    {"PauliX", PauliId::PauliId_X},
+    {"PauliY", PauliId::PauliId_Y},
+    {"PauliZ", PauliId::PauliId_Z}};
+
+const char InteropResultZeroAsChar = 0x0;
+const char InteropResultOneAsChar = 0x1;
+map<string, char> ResultAsCharMap{
+    {"0", InteropResultZeroAsChar},
+    {"Zero", InteropResultZeroAsChar},
+    {"1", InteropResultOneAsChar},
+    {"One", InteropResultOneAsChar}
+};
+
+template<typename T>
+unique_ptr<InteropArray> CreateInteropArray(vector<T>& v)
+{
+    unique_ptr<InteropArray> array(new InteropArray(v.size(), v.data()));
+    return array;
+}
+
+unique_ptr<InteropRange> CreateInteropRange(RangeTuple rangeTuple)
+{
+    unique_ptr<InteropRange> range(new InteropRange(rangeTuple));
+    return range;
+}
+
+template<typename T>
+void FreePointerVector(vector<T*>& v)
+{
+    for (auto p : v)
+    {
+        delete p;
+    }
+}
+
+char TranslatePauliToChar(PauliId& pauli)
+{
+    return static_cast<char>(pauli);
+}
+
+template<typename S, typename D>
+void TranslateVector(vector<S>& sourceVector, vector<D>& destinationVector, function<D(S&)> translationFunction)
+{
+    destinationVector.resize(sourceVector.size());
+    transform(sourceVector.begin(), sourceVector.end(), destinationVector.begin(), translationFunction);
+}
+
+InteropRange* TranslateRangeTupleToInteropRangePointer(RangeTuple& rangeTuple)
+{
+    InteropRange* range = new InteropRange(rangeTuple);
+    return range;
+}
+
+const char* TranslateStringToCharBuffer(string& s)
+{
+    return s.c_str();
+}
 
 int main(int argc, char* argv[])
 {
-    return 0;
+    CLI::App app("QIR Standalone Entry Point Inputs Reference");
+
+    // Initialize simulator.
+    unique_ptr<IRuntimeDriver> sim = CreateFullstateSimulator();
+    QirContextScope qirctx(sim.get(), false /*trackAllocatedObjects*/);
+
+    // Add the --simulation-output options.
+    // N.B. This option should be present in all standalone drivers.
+    string simulationOutputFile;
+    CLI::Option* simulationOutputFileOpt = app.add_option(
+        "--simulation-output", simulationOutputFile,
+        "File where the output produced during the simulation is written");
+    
+<# foreach (var arg in entryPointOperation.InteropArguments) {
+    WriteLine("");
+    WriteLine($"    {arg.CppVarType()} {arg.Name};"); 
+    if (arg.CppVarInitialValue() != null) { WriteLine($"    {arg.Name} = {arg.CppVarInitialValue()};"); }
+    WriteLine($"    app.add_option(\"{arg.CliOptionString()}\", {arg.Name}, \"{arg.CliDescription()}\")");
+    if (arg.TransformationType() == null) {
+        WriteLine("        ->required();");
+    } else {
+        WriteLine($"        ->required()->transform(CLI::CheckedTransformer({arg.TransformationType()}, CLI::ignore_case));");
+    }
+} #>
+
+    // With all the options added, parse arguments from the command line.
+    CLI11_PARSE(app, argc, argv);
+
+<# foreach (var arg in entryPointOperation.InteropArguments) {
```

Comment:
What I meant originally with this [comment](https://github.com/microsoft/qsharp-compiler/pull/929#discussion_r603590674) was something like the following:

For every argument there are two variables. One that holds the value parsed from the command line whose suffix would be _CliValue_, and one that holds the interop value to be passed to the entry point operation whose suffix would be _Interop_.

In the previous loop, the _CliValue_ variable is declared and used for the option as you are currently doing it:
```
WriteLine($"{arg.CppCliValueType()} {arg.CliValueVariableName()};");
```

In this loop, the _Interop_ variable is declared and the _CliValue_ is used to assign a value to the _Interop_ variable depending on the argument type. For example, for the PauliType it would be something like this:
```
WriteLine($"{arg.CppType()} {arg.InteropVariableName()};");
switch (arg.Type) {
    case DataType.PauliType:
          WriteLine($"{arg.InteropVariableName()} =TranslatePauliToChar({arg.CliValueVariableName()});");
          break;
```

This way, in the next loop, a switch is not needed and something simpler can be done:
```
Write($"{arg.InteropVariableName()}");
```

Reply:
I'll close the original comment and track the suggestion using this one.

Reply:
The one hiccup is in some cases there is another intermediate variable

Reply:
I think that's ok. It provides consistency on how the values for the entry-point operation are obtained from the command line and processed for interop with QIR. When the C++ source code is compiled, it's probably going to be optimized anyway.

Reply:
done